python/ray/tune/ray_trial_executor.py

```diff
@@ -0,0 +1,254 @@
+# coding: utf-8
+import os
+import traceback
+
+import time
+
+import ray
+from ray.tune.logger import NoopLogger
+
+from ray.tune.trial import Trial, Resources, Checkpoint
+from ray.tune.trial_executor import TrialExecutor
+
+class RayTrialExecutor(TrialExecutor):
+    """An implemention of TrialExecutor based on Ray."""
+
+    def __init__(self, queue_trials=False):
+        super(RayTrialExecutor, self).__init__(queue_trials)
+        self._running = {}
+        self._avail_resources = Resources(cpu=0, gpu=0)
+        self._committed_resources = Resources(cpu=0, gpu=0)
+        self._resources_initialized = False
+
+    def _setup_runner(self, trial):
+        cls = ray.remote(
+            num_cpus=trial.resources.cpu,
+            num_gpus=trial.resources.gpu)(trial._get_trainable_cls())
+
+        trial.init_logger()
+        remote_logdir = trial.logdir
+
+        def logger_creator(config):
+            # Set the working dir in the remote process, for user file writes
+            if not os.path.exists(remote_logdir):
+                os.makedirs(remote_logdir)
+            os.chdir(remote_logdir)
+            return NoopLogger(config, remote_logdir)
+
+        # Logging for trials is handled centrally by TrialRunner, so
+        # configure the remote runner to use a noop-logger.
+        return cls.remote(
+            config=trial.config, logger_creator=logger_creator)
+
+    def _add_running_trial(self, trial):
+        """get Ray future for one iteration of training"""
+
+        assert trial.status == Trial.RUNNING, trial.status
+        remote = trial.runner.train.remote()
+        self._running[remote] = trial
+
+    def _start_trial(self, trial, checkpoint=None):
+        trial.status = Trial.RUNNING
+        trial.runner = self._setup_runner(trial)
+        if self.restore(trial, checkpoint):
```

Comment:
_restore?

Reply:
Checkpoint mechanism seems not well designed. pbt.py PopulationBasedTraining calls save directly, so i set the function(save/restore) public.